cpp/serve/sampler.cc

```diff
@@ -391,20 +480,33 @@ class CPUSampler : public SamplerObj {
     CHECK_EQ(probs->ndim, 2);
     CHECK_EQ(probs->device.device_type, kDLCPU);
 
+    const int max_num_rejects = 6;
     int n = probs->shape[0];
-    std::vector<double> random_numbers;
+    std::vector<std::vector<double>> random_numbers;
     std::vector<int32_t> sampled_tokens;
     random_numbers.reserve(n);
     sampled_tokens.resize(n);
     for (int i = 0; i < n; ++i) {
-      random_numbers.push_back(rng_.GetRandomNumber());
+      std::vector<double> random_numbers_i;
+      for (int j = 0; j < max_num_rejects; ++j) {
+        random_numbers_i.push_back(rng_.GetRandomNumber());
+      }
+      random_numbers.push_back(random_numbers_i);
     }
 
     tvm::runtime::parallel_for_with_threading_backend(
         [&sampled_tokens, &probs, &generation_cfg, &random_numbers](int i) {
           // Sample top p from probability.
-          sampled_tokens[i] =
-              SampleTopPFromProb(probs, i, generation_cfg[i]->top_p, random_numbers[i]);
+          int reject_sampled_token =
+              RejectSampleTopPFromProb(probs, i, generation_cfg[i]->top_p, random_numbers[i]);
```

Comment:
combine the two sampling function together as sub functions functions of a global sample function that takes in prob and NDArray of uniform